### PR TITLE
Optimizes internal trace context related code, especially extra

### PR DESCRIPTION
**Pull request by adriancole**
After some trial and error with benchmarking, I found
`MutableTraceContext` is not enough a win unless we change api, which we
can't until Brave 6. Meanwhile, this optimizes as much as possible,
notably eliminating as many arraylist creations (from extra) as
possible.

Alternative to #755

### CODE REVIEW

**Comment by adriancole** [comment]
here are the results, btw


before:

```
Benchmark                                                                                                       Mode      Cnt      Score      Error   Units
TracerBenchmarks.joinWithSpanInScope:joinWithSpanInScope·p0.999                                               sample              21.216              us/op
TracerBenchmarks.joinWithSpanInScope:·gc.alloc.rate.norm                                                      sample       15    791.518 ±    1.755    B/op
TracerBenchmarks.joinWithSpanInScope_extra:joinWithSpanInScope_extra·p0.999                                   sample              23.872              us/op
TracerBenchmarks.joinWithSpanInScope_extra:·gc.alloc.rate.norm                                                sample       15    782.750 ±    6.044    B/op
TracerBenchmarks.joinWithSpanInScope_unsampled:joinWithSpanInScope_unsampled·p0.999                           sample               1.100              us/op
TracerBenchmarks.joinWithSpanInScope_unsampled:·gc.alloc.rate.norm                                            sample       15     40.008 ±    0.005    B/op
TracerBenchmarks.joinWithSpanInScope_unsampled_extra:joinWithSpanInScope_unsampled_extra·p0.999               sample              31.611              us/op
TracerBenchmarks.joinWithSpanInScope_unsampled_extra:·gc.alloc.rate.norm                                      sample       15     40.055 ±    0.004    B/op
TracerBenchmarks.newChildWithSpanInScope:newChildWithSpanInScope·p0.999                                       sample              28.770              us/op
TracerBenchmarks.newChildWithSpanInScope:·gc.alloc.rate.norm                                                  sample       15   1336.272 ±    0.096    B/op
TracerBenchmarks.newChildWithSpanInScope_extra:newChildWithSpanInScope_extra·p0.999                           sample              20.769              us/op
TracerBenchmarks.newChildWithSpanInScope_extra:·gc.alloc.rate.norm                                            sample       15   1712.572 ±   37.532    B/op
TracerBenchmarks.newChildWithSpanInScope_unsampled:newChildWithSpanInScope_unsampled·p0.999                   sample               1.682              us/op
TracerBenchmarks.newChildWithSpanInScope_unsampled:·gc.alloc.rate.norm                                        sample       15     96.010 ±    0.003    B/op
TracerBenchmarks.newChildWithSpanInScope_unsampled_extra:newChildWithSpanInScope_unsampled_extra·p0.999       sample              16.352              us/op
TracerBenchmarks.newChildWithSpanInScope_unsampled_extra:·gc.alloc.rate.norm                                  sample       15    424.346 ±    0.046    B/op
TracerBenchmarks.nextWithSpanInScope:nextWithSpanInScope·p0.999                                               sample              11.152              us/op
TracerBenchmarks.nextWithSpanInScope:·gc.alloc.rate.norm                                                      sample       15   1384.177 ±   37.565    B/op
TracerBenchmarks.nextWithSpanInScope_extra:nextWithSpanInScope_extra·p0.999                                   sample              14.320              us/op
TracerBenchmarks.nextWithSpanInScope_extra:·gc.alloc.rate.norm                                                sample       15   1760.513 ±    0.033    B/op
TracerBenchmarks.nextWithSpanInScope_unsampled:nextWithSpanInScope_unsampled·p0.999                           sample               1.756              us/op
TracerBenchmarks.nextWithSpanInScope_unsampled:·gc.alloc.rate.norm                                            sample       15     96.007 ±    0.001    B/op
TracerBenchmarks.nextWithSpanInScope_unsampled_extra:nextWithSpanInScope_unsampled_extra·p0.999               sample              13.182              us/op
TracerBenchmarks.nextWithSpanInScope_unsampled_extra:·gc.alloc.rate.norm                                      sample       15    565.759 ±    4.174    B/op
TracerBenchmarks.startScopedSpanWithParent:startScopedSpanWithParent·p0.999                                   sample              11.742              us/op
TracerBenchmarks.startScopedSpanWithParent:·gc.alloc.rate.norm                                                sample       15   1296.177 ±    0.031    B/op
TracerBenchmarks.startScopedSpanWithParent_extra:startScopedSpanWithParent_extra·p0.999                       sample              15.318              us/op
TracerBenchmarks.startScopedSpanWithParent_extra:·gc.alloc.rate.norm                                          sample       15   1672.493 ±   37.556    B/op
TracerBenchmarks.startScopedSpanWithParent_unsampled:startScopedSpanWithParent_unsampled·p0.999               sample               1.746              us/op
TracerBenchmarks.startScopedSpanWithParent_unsampled:·gc.alloc.rate.norm                                      sample       15    104.010 ±    0.006    B/op
TracerBenchmarks.startScopedSpanWithParent_unsampled_extra:startScopedSpanWithParent_unsampled_extra·p0.999   sample              17.600              us/op
TracerBenchmarks.startScopedSpanWithParent_unsampled_extra:·gc.alloc.rate.norm                                sample       15    440.389 ±   12.483    B/op
```


after:
```
Benchmark                                                                                                       Mode      Cnt      Score      Error   Units
TracerBenchmarks.joinWithSpanInScope:joinWithSpanInScope·p0.999                                               sample              23.616             us/op
TracerBenchmarks.joinWithSpanInScope:·gc.alloc.rate.norm                                                      sample       15    735.714 ±   4.159    B/op
TracerBenchmarks.joinWithSpanInScope_extra:joinWithSpanInScope_extra·p0.999                                   sample              23.392             us/op
TracerBenchmarks.joinWithSpanInScope_extra:·gc.alloc.rate.norm                                                sample       15    722.041 ±   3.148    B/op
TracerBenchmarks.joinWithSpanInScope_unsampled:joinWithSpanInScope_unsampled·p0.999                           sample               2.274             us/op
TracerBenchmarks.joinWithSpanInScope_unsampled:·gc.alloc.rate.norm                                            sample       15     40.009 ±   0.006    B/op
TracerBenchmarks.joinWithSpanInScope_unsampled_extra:joinWithSpanInScope_unsampled_extra·p0.999               sample              25.632             us/op
TracerBenchmarks.joinWithSpanInScope_unsampled_extra:·gc.alloc.rate.norm                                      sample       15     40.053 ±   0.004    B/op
TracerBenchmarks.newChildWithSpanInScope:newChildWithSpanInScope·p0.999                                       sample              11.216             us/op
TracerBenchmarks.newChildWithSpanInScope:·gc.alloc.rate.norm                                                  sample       15   1288.179 ±  75.129    B/op
TracerBenchmarks.newChildWithSpanInScope_extra:newChildWithSpanInScope_extra·p0.999                           sample              12.048             us/op
TracerBenchmarks.newChildWithSpanInScope_extra:·gc.alloc.rate.norm                                            sample       15   1512.306 ±  12.507    B/op
TracerBenchmarks.newChildWithSpanInScope_unsampled:newChildWithSpanInScope_unsampled·p0.999                   sample               0.779             us/op
TracerBenchmarks.newChildWithSpanInScope_unsampled:·gc.alloc.rate.norm                                        sample       15     96.007 ±   0.002    B/op
TracerBenchmarks.newChildWithSpanInScope_unsampled_extra:newChildWithSpanInScope_unsampled_extra·p0.999       sample               9.920             us/op
TracerBenchmarks.newChildWithSpanInScope_unsampled_extra:·gc.alloc.rate.norm                                  sample       15    336.123 ±   0.008    B/op
TracerBenchmarks.nextWithSpanInScope:nextWithSpanInScope·p0.999                                               sample              11.040             us/op
TracerBenchmarks.nextWithSpanInScope:·gc.alloc.rate.norm                                                      sample       15   1336.182 ±   0.012    B/op
TracerBenchmarks.nextWithSpanInScope_extra:nextWithSpanInScope_extra·p0.999                                   sample              11.808             us/op
TracerBenchmarks.nextWithSpanInScope_extra:·gc.alloc.rate.norm                                                sample       15   1544.300 ±  45.141    B/op
TracerBenchmarks.nextWithSpanInScope_unsampled:nextWithSpanInScope_unsampled·p0.999                           sample               0.816             us/op
TracerBenchmarks.nextWithSpanInScope_unsampled:·gc.alloc.rate.norm                                            sample       15     96.006 ±   0.001    B/op
TracerBenchmarks.nextWithSpanInScope_unsampled_extra:nextWithSpanInScope_unsampled_extra·p0.999               sample              16.866             us/op
TracerBenchmarks.nextWithSpanInScope_unsampled_extra:·gc.alloc.rate.norm                                      sample       15    448.328 ±  12.505    B/op
TracerBenchmarks.startScopedSpanWithParent:startScopedSpanWithParent·p0.999                                   sample               9.600             us/op
TracerBenchmarks.startScopedSpanWithParent:·gc.alloc.rate.norm                                                sample       15   1248.145 ±  37.564    B/op
TracerBenchmarks.startScopedSpanWithParent_extra:startScopedSpanWithParent_extra·p0.999                       sample              10.336             us/op
TracerBenchmarks.startScopedSpanWithParent_extra:·gc.alloc.rate.norm                                          sample       15   1496.261 ±  50.079    B/op
TracerBenchmarks.startScopedSpanWithParent_unsampled:startScopedSpanWithParent_unsampled·p0.999               sample               0.660             us/op
TracerBenchmarks.startScopedSpanWithParent_unsampled:·gc.alloc.rate.norm                                      sample       15    104.006 ±   0.001    B/op
TracerBenchmarks.startScopedSpanWithParent_unsampled_extra:startScopedSpanWithParent_unsampled_extra·p0.999   sample               5.975             us/op
TracerBenchmarks.startScopedSpanWithParent_unsampled_extra:·gc.alloc.rate.norm                                sample       15    352.117 ±  12.520    B/op
```

File: brave/src/main/java/brave/internal/recorder/PendingSpans.java [hunk 1]
```diff
@@ -75,23 +83,26 @@ public PendingSpan getOrCreate(TraceContext context) {
...
-    if (Boolean.TRUE.equals(context.shared())) {
-      TraceContext.Builder lookupContext = context.toBuilder().shared(false);
-      if (parentId != 0L) lookupContext.spanId(parentId);
-      parent = delegate.get(new LookupKey(lookupContext.build()));
-    } else if (parentId == 0L) {
-      parent = null; // root span was checked earlier
-    } else {
-      parent = delegate.get(new LookupKey(context.toBuilder().spanId(parentId).build()));
+    PendingSpan parent = null;
+    if (context.shared()) {
```
[the remark this answers is not in the document]
Comment (adriancole) [reply brave/src/main/java/brave/internal/recorder/PendingSpans.java]: done

File: brave/src/main/java/brave/Tracer.java [hunk 2]
```diff
@@ -182,7 +189,51 @@ public final Span joinSpan(TraceContext context) {
...
+      List<Object> extra
+  ) {
+    long nextId = nextId();
+    long parentId = 0L;
+    if (traceId == 0L) { // make a new trace ID
+      traceIdHigh = traceId128Bit ? Platform.get().nextTraceIdHigh() : 0L;
+      traceId = nextId;
+    } else { // try to set the parent from the extracted state
+      parentId = spanId;
+      flags &= ~FLAG_SHARED;
```
[the remark this answers is not in the document]
Comment (adriancole) [reply brave/src/main/java/brave/Tracer.java]: done

File: brave/src/main/java/brave/Tracer.java [hunk 3]
```diff
@@ -182,7 +189,51 @@ public final Span joinSpan(TraceContext context) {
...
+        parent.spanId(),
+        parent.extra()
+    );
+  }
+
+  TraceContext nextContext(
+      int flags,
+      long traceIdHigh,
+      long traceId,
+      long spanId,
```
[the remark this answers is not in the document]
Comment (adriancole) [reply brave/src/main/java/brave/Tracer.java]: done

File: brave/src/main/java/brave/internal/PropagationFieldsFactory.java [hunk 4]
```diff
@@ -14,41 +13,97 @@
...
-      fields = create((P) copyOfExtra.get(parentExtraIndex));
-      copyOfExtra.set(parentExtraIndex, fields);
-    } else if (thisExtraIndex != -1) { // merge with the parent
-      fields = ((P) copyOfExtra.get(thisExtraIndex));
-      P parent = (P) copyOfExtra.remove(parentExtraIndex); // ensures only one fields
-      fields.putAllIfAbsent(parent); // extracted wins vs parent
-    } else { // no fields were extracted and the parent also had no fields. create a new copy
-      fields = create();
-      copyOfExtra.add(fields);
+  static List<Object> ensureMutable(List<Object> list) {
```
[the remark this answers is not in the document]
Comment (adriancole) [reply brave/src/main/java/brave/internal/PropagationFieldsFactory.java]: done

File: brave/src/main/java/brave/internal/MapPropagationFields.java [hunk 5]
```diff
@@ -34,7 +34,9 @@ protected MapPropagationFields(MapPropagationFields parent) {
       if (values == null) {
         values = new LinkedHashMap<>();
         values.put(name, value);
-      } else if (!value.equals(values.get(name))) {
+      } else if (value.equals(values.get(name))) {
```
[the remark this answers is not in the document]
Comment (adriancole) [reply brave/src/main/java/brave/internal/MapPropagationFields.java]: done

File: brave/src/main/java/brave/internal/PredefinedPropagationFields.java [hunk 6]
```diff
@@ -50,8 +50,9 @@ public final void put(int index, String value) {
       if (elements == null) {
         elements = new String[fieldNames.length];
         elements[index] = value;
-      } else if (!value.equals(elements[index])) {
-        // this is the copy-on-write part
+      } else if (value.equals(elements[index])) {
+        return;
```
[the remark this answers is not in the document]
Comment (adriancole) [reply brave/src/main/java/brave/internal/PredefinedPropagationFields.java]: done

File: brave/src/main/java/brave/internal/PropagationFields.java [hunk 7]
```diff
@@ -20,14 +19,15 @@
   public abstract Map<String, String> toMap();
 
   /** Fields are extracted before a context is created. We need to lazy set the context */
-  final boolean tryAssociate(TraceContext newContext) {
+  final boolean tryToClaim(long traceId, long spanId) {
```
[the remark this answers is not in the document]
Comment (adriancole) [reply brave/src/main/java/brave/internal/PropagationFields.java]: still needed for edge cases

File: brave/src/main/java/brave/internal/recorder/PendingSpans.java [hunk 8]
```diff
@@ -27,6 +27,8 @@
  * https://github.com/raphw/weak-lock-free/blob/master/src/main/java/com/blogspot/mydailyjava/weaklockfree/WeakConcurrentMap.java
  */
 public final class PendingSpans extends ReferenceQueue<TraceContext> {
+  static final int FLAG_SHARED = 1 << 4;
```
[the remark this answers is not in the document]
Comment (adriancole) [reply brave/src/main/java/brave/internal/recorder/PendingSpans.java]: done

File: brave/src/main/java/brave/propagation/TraceContextOrSamplingFlags.java [hunk 9]
```diff
@@ -47,14 +51,20 @@ public static Builder newBuilder(){
...
       case 1:
-        return new TraceContextOrSamplingFlags(type, ((TraceContext) value).toBuilder()
-            .sampled(sampled).build(), extra);
+        return new TraceContextOrSamplingFlags(
+            type, contextWithSampled((TraceContext) value, sampled), extra);
       case 2:
-        return new TraceContextOrSamplingFlags(type, ((TraceIdContext) value).toBuilder()
-            .sampled(sampled).build(), extra);
+        return new TraceContextOrSamplingFlags(
+            type, ((TraceIdContext) value).toBuilder().sampled(sampled).build(), extra);
```
[the remark this answers is not in the document]
Comment (adriancole) [reply brave/src/main/java/brave/propagation/TraceContextOrSamplingFlags.java]: done

File: brave/src/main/java/brave/propagation/TraceIdContext.java [hunk 10]
```diff
@@ -30,7 +26,12 @@ public long traceId() {
   }
 
   public Builder toBuilder() {
-    return new Builder(this);
+    Builder result = new Builder();
+    result.traceIdHigh(traceIdHigh);
```
[the remark this answers is not in the document]
Comment (adriancole) [reply brave/src/main/java/brave/propagation/TraceIdContext.java]: done

**Comment by adriancole** [comment]
running bench after latest. will merge in the morning and move on unless someone screams

**Comment by adriancole** [comment]
using flags directly is measurably quicker vs helpers. I will revert to using them

**Comment by yurishkuro** [comment]
You would expect jvm to inline the helpers.

**Comment by adriancole** [comment]
@yurishkuro we can revisit later as it is internal code. np

**Comment by adriancole** [comment]
by revisit, what I mean is that places which are using bitwise to quick check on something.. someone can run the bench only changing this specific line. The problem is that the benchmarks take a very long time to execute (in sampled mode), line 20m to run just TracerBenchmarks. It is possible some lines can be same as if using dispatch, but anyway can't personally afford to spend too much more time on this right now, so good enough basically.

**Pull request merged**